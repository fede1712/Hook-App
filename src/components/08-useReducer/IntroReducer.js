const initialState = [{
    id: 1,
    todo: 'Comprar leche',
    done: false
}]
const newTodo = {
    id: 2,
    todo: 'Comprar pan',
    done: false
}

const action = {
    type: 'add',
    payload: newTodo
}

const todoReducer = (state = initialState, action) => {
    if (action?.type === 'add') {
        return [...state, action.payload]
    }

    return state
}

let todos = todoReducer()


todos = todoReducer(todos, action)





console.log(todos)