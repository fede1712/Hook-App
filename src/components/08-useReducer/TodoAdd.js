import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handleAddTodo }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    })
    const handleSubmit = (e) => {
        e.preventDefault()
        if (description.trim().length < 1) {
            return
        }

        handleAddTodo({
            id: new Date().getTime(),
            description: description,
            done: false
        })
        reset()
    }

    return (
        <>
            <h4>Add TODO</h4>
            <hr />
            <form onSubmit={handleSubmit} >
                <input
                    className='form-control'
                    type='text'
                    name='description'
                    placeholder='To do item'
                    autoComplete='off'
                    value={description}
                    onChange={handleInputChange}
                />
                <button className='btn btn-outline-primary mt-1 btn-block' type='submit'>Add</button>
            </form>

        </>
    )
}
