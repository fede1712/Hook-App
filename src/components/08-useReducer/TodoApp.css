body {
    padding: 70px;
}

p {
    cursor: pointer;
    margin: 0;
}

.list-group-item {
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.complete {
    font-weight: bold;
    text-decoration: line-through;
    color: green;
}

.not-complete {
    color: red;
    font-weight: bold;
}