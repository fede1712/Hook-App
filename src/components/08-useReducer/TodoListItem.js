import React from 'react'

export const TodoListItem = ({ elm, handleDelete, handleToggle }) => {
    return (
        <li key={elm.id} className='list-group-item'>
            <p className={`${elm.done ? 'complete' : 'not-complete'}`} onClick={() => handleToggle(elm.id)}>{elm.description}</p>
            <button className='btn btn-outline-danger' onClick={() => handleDelete(elm.id)}>Delete</button>
        </li>
    )
}
