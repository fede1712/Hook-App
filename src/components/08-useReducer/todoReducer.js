export const todoReducer = (state = [], action) => {

    switch (action.type) {
        case 'add':
            return [...state, action.payload]

        case 'delete':
            return state.filter(elm => elm.id !== action.payload)
        case 'done':
            return state.map(elm => (elm.id === action.payload ? { ...elm, done: !elm.done } : elm))
        default:
            return state;
    }

}