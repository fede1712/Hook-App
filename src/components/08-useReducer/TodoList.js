import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul className='list-group list-group-flush'>
            {todos.map(elm => (
                <TodoListItem key={elm.id} elm={elm} handleDelete={handleDelete} handleToggle={handleToggle} />
            )
            )}
        </ul>
    )
}
