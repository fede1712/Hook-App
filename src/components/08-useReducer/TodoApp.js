import React from 'react'
import { useReducer } from 'react'
import { todoReducer } from './todoReducer'
import './TodoApp.css'

const initialState = [{
    id: new Date().getTime(),
    description: 'Aprender React',
    done: false
}]

export const TodoApp = () => {

    const [todos] = useReducer(todoReducer, initialState)


    return (
        <div>
            <h1>TODO App ({todos.length})</h1>
            <hr />
            <ul className='list-group list-group-flush'>
                {todos.map(elm => (
                    <>
                        <li key={elm.id} className='list-group-item'>
                            <p className='text-center'>{elm.description}</p>
                        </li>
                        <button className='btn btn-outline-danger'>Delete</button>
                    </>
                )
                )}
            </ul>
        </div>
    )
}
