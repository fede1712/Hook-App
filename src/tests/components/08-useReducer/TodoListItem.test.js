import React from 'react'
import { shallow } from 'enzyme'
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem'
import { demoTodos } from '../../fixtures/demoTodos'

describe('Tests in <TodoListItem />', () => {
    const handleDelete = jest.fn()
    const handleToggle = jest.fn()

    const wrapper = shallow(<TodoListItem elm={demoTodos[0]} handleDelete={handleDelete} handleToggle={handleToggle} />)

    test('Should render <TodoListItem /> correctly', () => {

        expect(wrapper).toMatchSnapshot()

    })
    test('Should call delete function', () => {

        wrapper.find('button').simulate('click')

        expect(handleDelete).toHaveBeenCalled()
        expect(handleDelete).toHaveBeenCalledWith(demoTodos[0].id)

    })

    test('Should call handleToggle function', () => {

        wrapper.find('p').simulate('click')

        expect(handleToggle).toHaveBeenCalled()
        expect(handleToggle).toHaveBeenCalledWith(demoTodos[0].id)

    })

    test('should have complete className if TODO.id = done', () => {

        const todo = demoTodos[0]

        todo.done = true

        const wrapper = shallow(<TodoListItem elm={todo} />)

        expect(wrapper.find('p').hasClass('complete')).toBe(true)



    })

})