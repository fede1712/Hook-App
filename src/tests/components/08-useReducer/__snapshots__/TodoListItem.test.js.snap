// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoListItem /> Should render <TodoListItem /> correctly 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="not-complete"
    onClick={[Function]}
  />
  <button
    className="btn btn-outline-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
